storage/indexer/src/db_tailer.rs

```diff
@@ -0,0 +1,302 @@
+    pub db: Arc<DB>,
+    pub main_db_reader: Arc<dyn DbReader>,
+    batch_size: usize,
+}
+
+impl DBTailer {
+    pub fn new(db: Arc<DB>, db_reader: Arc<dyn DbReader>, config: &IndexDBTailerConfig) -> Self {
+        Self {
+            db,
+            main_db_reader: db_reader,
+            batch_size: config.batch_size,
+        }
+    }
+
+    pub fn get_persisted_version(&self) -> Version {
+        // read the latest key from the db
+        let mut rev_iter_res = self
+            .db
+            .rev_iter::<TailerMetadataSchema>(Default::default())
+            .expect("Cannot create db tailer metadata iterator");
+        rev_iter_res
+            .next()
+            .map(|res| res.map_or(0, |(version, _)| version))
+            .unwrap_or_default()
+    }
+
+    pub fn process_a_batch(&self, start_version: Option<Version>) -> Result<Version> {
+        let mut version = start_version.unwrap_or(0);
+        let db_iter: Box<
+            dyn Iterator<
+                Item = std::prelude::v1::Result<
+                    (aptos_types::transaction::Transaction, Vec<ContractEvent>),
+                    AptosDbError,
+                >,
+            >,
+        > = self
+            .main_db_reader
+            .get_db_backup_iter(version, self.batch_size)
+            .expect("Cannot create db tailer iterator");
+        let batch = SchemaBatch::new();
+        let metadata_batch = SchemaBatch::new();
+        db_iter.for_each(|res| {
+            res.map(|(txn, events)| {
+                if let Some(txn) = txn.try_as_signed_user_txn() {
+                    batch
+                        .put::<TransactionByAccountSchema>(
+                            &(txn.sender(), txn.sequence_number()),
+                            &version,
+                        )
+                        .expect("Failed to put txn to db tailer batch");
+
+                    events.iter().enumerate().for_each(|(idx, event)| {
+                        if let ContractEvent::V1(v1) = event {
+                            batch
+                                .put::<EventByKeySchema>(
+                                    &(*v1.key(), v1.sequence_number()),
+                                    &(version, idx as u64),
+                                )
+                                .expect("Failed to event by key to db tailer batch");
+                            batch
+                                .put::<EventByVersionSchema>(
+                                    &(*v1.key(), version, v1.sequence_number()),
+                                    &(idx as u64),
+                                )
+                                .expect("Failed to event by version to db tailer batch");
+                        }
+                    });
+                }
+                version += 1;
+            })
+            .expect("Failed to iterate db tailer iterator");
+        });
+        // write to index db
+        self.db.write_schemas(batch)?;
```

Comment:
better to write data and metadata in the same txn since they are only touching one database?